### PR TITLE
Vector drawables

### PR DESCRIPTION
**Pull request by florina-muntenescu**

### CODE REVIEW

File: _posts/blog/2016-10-08-vector_drawables_optimisation.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+categories: blog
+author: florina_muntenescu
+excerpt: You have replaced all image resources in your app with VectorDrawables - your APK is smaller and your images look better. But are you really using them correctly? Here are three mistakes that you might be making and how to fix them, right now. Because &#35;PERFMATTERS.
+tags: [Android, UI, Vector Drawables, perfmatters]
+image:
+date: 2016-10-08T00:39:55-04:00
+---
+
+
+While some mobile platforms have been supporting vector graphics for a while, Android only began doing this natively starting with API Level 21 and with the help of the Support Library 23.2.0 for pre-Lollipop devices. By replacing you PNG image resources with VectorDrawables, your APK size decreases considerably and most of all, your images look good, independent of the resolution of the device used.
```
Comment (ghost) [inline _posts/blog/2016-10-08-vector_drawables_optimisation.md]: independent of the used resolution.

File: _posts/blog/2016-10-08-vector_drawables_optimisation.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+
+The system will try to redraw your Activity every 16ms so it can reach the <a href="https://www.youtube.com/watch?v=CaMTIgxCSqU">60fps</a> target. So this means that the duration of every `onDraw` method is extremely important.
+
+
+When used incorrectly, VectorDrawables can affect the performance of your app, because the drawing can take a long time. Here are three mistakes that you might be making, but can be solved easily, to ensure that you are really improving the performance of your app, with the help of VectorDrawables.
```
Comment (ghost) [inline _posts/blog/2016-10-08-vector_drawables_optimisation.md]: Here are three pitfalls that can be easily avoided....

File: _posts/blog/2016-10-08-vector_drawables_optimisation.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+
+## Understanding The Internals Of VectorDrawables
+
+
+Vector graphics use geometrical shapes to describe graphical elements. The vector graphics are rendered at runtime. The automatic rendering at pixel density gives smoothness to the graphics, regardless of the device capabilities. So your images won’t be downscaled or upscaled, looking stretched or pixelated but they will be always perfectly drawn for your screen size.
+Vector drawables allow the representation of images (e.g. icons, UI elements) based on XML vector graphics.
```
Comment (ghost) [inline _posts/blog/2016-10-08-vector_drawables_optimisation.md]: either delete the last sentence or put it at the beginning of the paragraph - it does not provide new information

File: _posts/blog/2016-10-08-vector_drawables_optimisation.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+</center>
+
+
+Using vector data instead of raster image data, the number and the size of resources added to the project decreases, since now only one resource per resolution is needed, and therefore also the APK size.
+
+
+Although the XML file containing the vector drawable is usually smaller than the PNG version, the vector drawables come with a computational overhead at runtime, which may be an issue for more complex graphical elements. When vector drawables are drawn for the first time, a cached bitmap is created in order to optimize the re-drawing performance. This cache is re-used as long as the width and the height of the image that needs to be drawn is the same. If a VectorDrawable is used for multiple sizes, a new Bitmap will be created every time.
+
+
+Compared to raster images, drawing VectorDrawables will take more time for the first rendering. But then, as long as the size of the image doesn't change, the next drawings of VectorDrawables will take a similar amount of time,  just like drawing raster images.
```
Comment (ghost) [inline _posts/blog/2016-10-08-vector_drawables_optimisation.md]: But then, as long as the size of the image doesn't change, the next drawings of VectorDrawables will take as much time as drawing raster images.

File: _posts/blog/2016-10-08-vector_drawables_optimisation.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+
+
+
+## Multiple Sizes, Multiple Renderings
+
+
+Let's say that you need to display an image that needs to fill the height of the screen but keep the aspect ration when rotating the device from portrait to landscape.
```
Comment (ghost) [inline _posts/blog/2016-10-08-vector_drawables_optimisation.md]: aspect ratio

File: _posts/blog/2016-10-08-vector_drawables_optimisation.md [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+<center>
+<picture>
+	<a href="/images/blog/vector_drawables/phone_portrait.png"><img src="/images/blog/vector_drawables/phone_portrait.png" alt="VectorDrawable vs PNG" height="350"></a>
+	<a href="/images/blog/vector_drawables/phone_landscape.png"><img src="/images/blog/vector_drawables/phone_landscape.png" alt="VectorDrawable vs PNG" width="350"></a>
+	<figcaption>Changing the size of the image on orientation</figcaption>
+</picture>
+</center>
+
+
+**VectorDrawables use a Bitmap cache that gets recreated when the size changes**, so for example when you rotate your device from portrait to landscape. After the first rendering, the cached bitmap will be used. This means that you end up spending a lot of time on the first rendering of the VectorDrawable, at every rotation of the screen.
```
Comment (ghost) [inline _posts/blog/2016-10-08-vector_drawables_optimisation.md]: could you rephrase the end of the sentence? it is a bit confusing - it sounds as the calculation would be done with every rotation -

File: _posts/blog/2016-10-08-vector_drawables_optimisation.md [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+The best approach, in this case,is to **create one VectorDrawable for portrait and one for landscape**. Like this, time will be spent only for the first rendering of the two images and the corresponding bitmap caches will be used from there on for every rotation.
```
Comment (ghost) [inline _posts/blog/2016-10-08-vector_drawables_optimisation.md]: ahh the next sentence explains it - so there is just one cached image?

File: _posts/blog/2016-10-08-vector_drawables_optimisation.md [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+
+Let's test this! The size of the view changes when changing the orientation. When rotating the device, drawing the vector drawable takes, in average, 15.50ms in portrait and 7.80ms in landscape, where the view is smaller.
```
Comment (ghost) [inline _posts/blog/2016-10-08-vector_drawables_optimisation.md]: I wouldn't use absolute numbers - isn't that dependent on the svg and the performance of the phone and so on?

Comment (florina-muntenescu) [reply _posts/blog/2016-10-08-vector_drawables_optimisation.md]: yes it is - that's why I'm mentioning that later in a note below, in the test section.
